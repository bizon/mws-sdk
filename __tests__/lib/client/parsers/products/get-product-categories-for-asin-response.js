const parseXml = require('../../../../../lib/client/parsers')
const parseGetProductCategoriesForAsinResponse = require('../../../../../lib/client/parsers/products/get-product-categories-for-asin-response')

describe('lib.client.parsers.products.get-product-categories-for-asin-response', () => {
  it('should parse the GetProductCategoriesForASINResponse example response from MWS doc', () => {
    const doc = parseXml(
      `<?xml version="1.0"?>
      <GetProductCategoriesForASINResponse xmlns="http://mws.amazonservices.com/schema/Products/2011-10-01">
        <GetProductCategoriesForASINResult>
          <Self>
            <ProductCategoryId>2420095011</ProductCategoryId>
            <ProductCategoryName>Compression Shorts</ProductCategoryName>
            <Parent>
              <ProductCategoryId>2419332011</ProductCategoryId>
              <ProductCategoryName>Men</ProductCategoryName>
              <Parent>
                <ProductCategoryId>2371051011</ProductCategoryId>
                <ProductCategoryName>Clothing</ProductCategoryName>
                <Parent>
                  <ProductCategoryId>3403201</ProductCategoryId>
                  <ProductCategoryName>Bikes &#x26; Accessories</ProductCategoryName>
                  <Parent>
                    <ProductCategoryId>2232464011</ProductCategoryId>
                    <ProductCategoryName>Bikes &#x26; Scooters</ProductCategoryName>
                    <Parent>
                      <ProductCategoryId>3375301</ProductCategoryId>
                      <ProductCategoryName>Categories</ProductCategoryName>
                      <Parent>
                        <ProductCategoryId>3375251</ProductCategoryId>
                        <ProductCategoryName>Categories</ProductCategoryName>
                      </Parent>
                    </Parent>
                  </Parent>
                </Parent>
              </Parent>
            </Parent>
          </Self>
        </GetProductCategoriesForASINResult>
        <ResponseMetadata>
          <RequestId>fbce5b62-67cc-4ab8-86f3-EXAMPLE22e4e</RequestId>
        </ResponseMetadata>
      </GetProductCategoriesForASINResponse>`,
    )

    const res = parseGetProductCategoriesForAsinResponse(
      '/products:GetProductCategoriesForASINResponse',
      doc,
    )

    expect(res).toMatchSnapshot()
  })

  it('should throw if there are too many levels of category recursion', () => {
    const doc = parseXml(
      `<?xml version="1.0"?>
      <GetProductCategoriesForASINResponse xmlns="http://mws.amazonservices.com/schema/Products/2011-10-01">
        <GetProductCategoriesForASINResult>
          <Self>
            <ProductCategoryId>1</ProductCategoryId>
            <ProductCategoryName>1</ProductCategoryName>
            <Parent>
              <ProductCategoryId>2</ProductCategoryId>
              <ProductCategoryName>2</ProductCategoryName>
              <Parent>
                <ProductCategoryId>3</ProductCategoryId>
                <ProductCategoryName>3</ProductCategoryName>
                <Parent>
                  <ProductCategoryId>4</ProductCategoryId>
                  <ProductCategoryName>4</ProductCategoryName>
                  <Parent>
                    <ProductCategoryId>5</ProductCategoryId>
                    <ProductCategoryName>5</ProductCategoryName>
                    <Parent>
                      <ProductCategoryId>6</ProductCategoryId>
                      <ProductCategoryName>6</ProductCategoryName>
                      <Parent>
                        <ProductCategoryId>7</ProductCategoryId>
                        <ProductCategoryName>7</ProductCategoryName>
                        <Parent>
                          <ProductCategoryId>8</ProductCategoryId>
                          <ProductCategoryName>8</ProductCategoryName>
                          <Parent>
                            <ProductCategoryId>9</ProductCategoryId>
                            <ProductCategoryName>9</ProductCategoryName>
                            <Parent>
                              <ProductCategoryId>10</ProductCategoryId>
                              <ProductCategoryName>10</ProductCategoryName>
                              <Parent>
                                <ProductCategoryId>11</ProductCategoryId>
                                <ProductCategoryName>11</ProductCategoryName>
                              </Parent>
                            </Parent>
                          </Parent>
                        </Parent>
                      </Parent>
                    </Parent>
                  </Parent>
                </Parent>
              </Parent>
            </Parent>
          </Self>
        </GetProductCategoriesForASINResult>
        <ResponseMetadata>
          <RequestId>e058aabd-b4c3-48ba-9bfa-EXAMPLE9a267</RequestId>
        </ResponseMetadata>
      </GetProductCategoriesForASINResponse>`,
    )

    expect(
      () => parseGetProductCategoriesForAsinResponse(
        '/products:GetProductCategoriesForASINResponse',
        doc,
      ),
    ).toThrow('Failed parsing product category: more than 10 levels of depth')
  })
})
