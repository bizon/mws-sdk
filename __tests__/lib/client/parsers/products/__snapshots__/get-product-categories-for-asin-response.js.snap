// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib.client.parsers.products.get-product-categories-for-asin-response should parse the GetProductCategoriesForASINResponse example response from MWS doc 1`] = `
Object {
  "getProductCategoriesForAsinResult": Object {
    "categories": Array [
      Object {
        "parent": Object {
          "parent": Object {
            "parent": Object {
              "parent": Object {
                "parent": Object {
                  "parent": Object {
                    "parent": null,
                    "productCategoryId": "3375251",
                    "productCategoryName": "Categories",
                  },
                  "productCategoryId": "3375301",
                  "productCategoryName": "Categories",
                },
                "productCategoryId": "2232464011",
                "productCategoryName": "Bikes & Scooters",
              },
              "productCategoryId": "3403201",
              "productCategoryName": "Bikes & Accessories",
            },
            "productCategoryId": "2371051011",
            "productCategoryName": "Clothing",
          },
          "productCategoryId": "2419332011",
          "productCategoryName": "Men",
        },
        "productCategoryId": "2420095011",
        "productCategoryName": "Compression Shorts",
      },
    ],
  },
  "responseMetadata": Object {
    "requestId": "fbce5b62-67cc-4ab8-86f3-EXAMPLE22e4e",
  },
}
`;
